File: finalFacebook/Sqrt(x).java
public class Solution {
    public int mySqrt(int x) {
        if (x < 0) {
            return -1;
        }
        long start = 1;
        long end = x;
        while (start + 1 < end) {
            long mid = start + (end - start) / 2;
            if (mid * mid == x) {
                return (int)mid;
            } else if (mid * mid > x) {
                end = mid;
            } else {
                start = mid;
            }
        }
        if (end * end <= x) {
            return (int)end;
        }
        return (int)start;
    }
}

File: Python/binarysearch/Sqrt(x).py
class Solution:
    def mySqrt(self, x: int) -> int:
        if x==1:
            return 1
        l=1
        r=x
        while(l<=r):
            mid =l+(r-l)//2
            if mid**2>x:
                r=mid-1
            if mid**2<x:
                l=mid+1
            if mid**2==x:
                return mid
        return r
#sol2
from math import e, log
class Solution:
    def mySqrt(self, x):
        if x < 2:
            return x
        
        left = int(e**(0.5 * log(x)))
        right = left + 1
        return left if right * right > x else right